datafusion/physical-plan/src/coalescer/mod.rs

```diff
@@ -0,0 +1,204 @@
+/// `target_batch_size` rows.
+///
+/// 3. This structure also handles other optimizations such as a
+/// combined filter/coalesce operation.
+#[derive(Debug)]
+pub struct BatchCoalescer {
+    /// The input schema
+    schema: SchemaRef,
+    /// Minimum number of rows for coalesces batches
+    target_batch_size: usize,
+    /// Buffered batches
+    buffer: Vec<RecordBatch>,
+    /// Buffered row count
+    buffered_rows: usize,
+}
+
+impl BatchCoalescer {
+    /// Create a new BatchCoalescer that produces batches of at least `target_batch_size` rows
+    pub fn new(schema: SchemaRef, target_batch_size: usize) -> Self {
+        Self {
+            schema,
+            target_batch_size,
+            buffer: vec![],
+            buffered_rows: 0,
+        }
+    }
+
+    /// Return the schema of the output batches
+    pub fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema)
+    }
+
+    /// Add a batch to the coalescer, returning a batch if the target batch size is reached
+    pub fn push_batch(&mut self, batch: RecordBatch) -> Result<Option<RecordBatch>> {
+        if batch.num_rows() >= self.target_batch_size && self.buffer.is_empty() {
+            return Ok(Some(batch));
+        }
+        // discard empty batches
+        if batch.num_rows() == 0 {
+            return Ok(None);
+        }
+        // add to the buffered batches
+        self.buffered_rows += batch.num_rows();
+        self.buffer.push(batch);
+        // check to see if we have enough batches yet
+        let batch = if self.buffered_rows >= self.target_batch_size {
+            // combine the batches and return
+            let batch = concat_batches(&self.schema, &self.buffer)?;
+            // reset buffer state
+            self.buffer.clear();
+            self.buffered_rows = 0;
+            // return batch
+            Some(batch)
+        } else {
+            None
+        };
+        Ok(batch)
+    }
+
+    /// Push the rows in `batch` that matching the `predicate` (i. e. where the values are true).
+    pub fn push_batch_with_filter(
+        &mut self,
+        batch: &RecordBatch,
+        predicate: &BooleanArray,
+    ) -> Result<Option<RecordBatch>> {
+        let mut filter_builder = FilterBuilder::new(predicate);
+        if batch.num_columns() > 1 {
+            // Only optimize if filtering more than one column
+            filter_builder = filter_builder.optimize();
+        }
+        let filter = filter_builder.build();
+
+        // TODO the idea is here is to iteratively build up the filtered batch
+        // and push it when it reaches the target size rather than buffering directly
+        let filtered_arrays = batch
+            .columns()
+            .iter()
+            .map(|a| filter_array(a, &filter))
+            .collect::<Result<Vec<_>, _>>()?;
+        let options = RecordBatchOptions::default().with_row_count(Some(filter.count()));
+        let filtered_batch =
```

Comment:
The key goal is to avoid this batch materialization. I'll keep hacking on it tomorrow / later